TestIsInferGroupOutdated failed

```
=== RUN   TestIsInferGroupOutdated/revision_not_match
=== RUN   TestIsInferGroupOutdated/revision_match
    /root/go/src/matrixinfer.ai/matrixinfer/pkg/infer-controller/controller/modelinfer_controller_test.go:109: expected false when revision matches
--- FAIL: TestIsInferGroupOutdated (0.00s)
    --- PASS: TestIsInferGroupOutdated/no_pods (0.00s)
    --- PASS: TestIsInferGroupOutdated/no_revision_label (0.00s)
    --- PASS: TestIsInferGroupOutdated/revision_not_match (0.00s)
    --- FAIL: TestIsInferGroupOutdated/revision_match (0.00s)
FAIL
FAIL	matrixinfer.ai/matrixinfer/pkg/infer-controller/controller	0.016s
```

--- a/pkg/infer-controller/controller/modelinfer_controller.go
+++ b/pkg/infer-controller/controller/modelinfer_controller.go
@@ -85,10 +85,7 @@ func NewModelInferController(kubeClientSet kubernetes.Interface, modelInferClien
 	modelInferInformerFactory := informersv1alpha1.NewSharedInformerFactory(modelInferClient, 0)
 	modelInferInformer := modelInferInformerFactory.Workload().V1alpha1().ModelInfers()
 
-	store, err := datastore.New()
-	if err != nil {
-		klog.Fatal("Unable to create data store")
-	}
+	store := datastore.New()
 
 	c := &ModelInferController{
 		kubeClientSet:       kubeClientSet,

--- a/pkg/infer-controller/controller/modelinfer_controller_test.go
+++ b/pkg/infer-controller/controller/modelinfer_controller_test.go
@@ -15,98 +15,215 @@ package controller
 
 import (
 	"context"
+	"fmt"
 	"testing"
+	"time"
 
+	"github.com/stretchr/testify/assert"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/client-go/informers"
-	"k8s.io/client-go/kubernetes/fake"
+	kubefake "k8s.io/client-go/kubernetes/fake"
 
-	"matrixinfer.ai/matrixinfer/pkg/apis/workload/v1alpha1"
+	workloadv1alpha1 "matrixinfer.ai/matrixinfer/pkg/apis/workload/v1alpha1"
 	"matrixinfer.ai/matrixinfer/pkg/infer-controller/datastore"
+	"matrixinfer.ai/matrixinfer/pkg/infer-controller/utils"
 )
 
-func newTestControllerWithPods(pods []*corev1.Pod, namespace string) *ModelInferController {
-	client := fake.NewSimpleClientset()
-	informerFactory := informers.NewSharedInformerFactory(client, 0)
-	podInformer := informerFactory.Core().V1().Pods()
-
-	// Create pods first
-	for _, pod := range pods {
-		_, _ = client.CoreV1().Pods(namespace).Create(context.TODO(), pod, metav1.CreateOptions{})
-	}
-	// Start informer and synchronize
-	stopCh := make(chan struct{})
-	defer close(stopCh)
-	informerFactory.Start(stopCh)
-	informerFactory.WaitForCacheSync(stopCh)
-
-	return &ModelInferController{
-		podsLister: podInformer.Lister(),
-	}
+type podSpec struct {
+	name   string
+	labels map[string]string
 }
 
 func TestIsInferGroupOutdated(t *testing.T) {
 	ns := "test-ns"
 	groupName := "test-group"
 	group := datastore.InferGroup{Name: groupName}
-	mi := &v1alpha1.ModelInfer{ObjectMeta: metav1.ObjectMeta{Namespace: ns}}
+	mi := &workloadv1alpha1.ModelInfer{ObjectMeta: metav1.ObjectMeta{Namespace: ns}}
 	newHash := "hash123"
 
-	t.Run("no pods", func(t *testing.T) {
-		c := newTestControllerWithPods([]*corev1.Pod{}, ns)
-		if !c.isInferGroupOutdated(group, mi.Namespace, newHash) {
-			t.Errorf("expected true when no pods")
-		}
-	})
+	kubeClient := kubefake.NewSimpleClientset()
+	informerFactory := informers.NewSharedInformerFactory(kubeClient, 0)
+	podInformer := informerFactory.Core().V1().Pods()
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+	informerFactory.Start(stopCh)
+	informerFactory.WaitForCacheSync(stopCh)
 
-	t.Run("no revision label", func(t *testing.T) {
+	c := &ModelInferController{
+		podsLister: podInformer.Lister(),
+	}
+
+	cases := []struct {
+		name string
+		pods []podSpec
+		want bool
+	}{
+		{
+			name: "no pods",
+			pods: nil,
+			want: true,
+		},
+		{
+			name: "no revision label",
+			pods: []podSpec{
+				{name: "pod1", labels: map[string]string{workloadv1alpha1.GroupNameLabelKey: groupName}},
+			},
+			want: true,
+		},
+		{
+			name: "revision not match",
+			pods: []podSpec{
+				{name: "pod2", labels: map[string]string{workloadv1alpha1.GroupNameLabelKey: groupName, workloadv1alpha1.RevisionLabelKey: "oldhash"}},
+			},
+			want: true,
+		},
+		{
+			name: "revision match",
+			pods: []podSpec{
+				{name: "pod3", labels: map[string]string{workloadv1alpha1.GroupNameLabelKey: groupName, workloadv1alpha1.RevisionLabelKey: newHash}},
+			},
+			want: false,
+		},
+	}
+
+	indexer := podInformer.Informer().GetIndexer()
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			// clean indexer
+			for _, obj := range indexer.List() {
+				err := indexer.Delete(obj)
+				assert.NoError(t, err)
+			}
+			for _, p := range tc.pods {
+				pod := &corev1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						Namespace: ns,
+						Name:      p.name,
+						Labels:    p.labels,
+					},
+				}
+				err := indexer.Add(pod)
+				assert.NoError(t, err)
+			}
+			got := c.isInferGroupOutdated(group, mi.Namespace, newHash)
+			assert.Equal(t, tc.want, got)
+		})
+	}
+}
+
+func TestCheckInferGroupReady(t *testing.T) {
+	ns := "default"
+	groupName := "test-group"
+	newHash := "hash123"
+
+	kubeClient := kubefake.NewSimpleClientset()
+	kubeInformerFactory := informers.NewSharedInformerFactory(kubeClient, 0)
+	podInformer := kubeInformerFactory.Core().V1().Pods()
+	store := datastore.New()
+	// build controller
+	controller := &ModelInferController{
+		podsLister: podInformer.Lister(),
+		store:      store,
+	}
+	stop := make(chan struct{})
+	defer close(stop)
+	kubeInformerFactory.Start(stop)
+	kubeInformerFactory.WaitForCacheSync(stop)
+
+	// build ModelInfer
+	var expectedPodNum int32 = 2
+	mi := &workloadv1alpha1.ModelInfer{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      "test-mi",
+		},
+		Spec: workloadv1alpha1.ModelInferSpec{
+			Template: workloadv1alpha1.InferGroup{
+				Roles: []workloadv1alpha1.Role{
+					{
+						Replicas: &expectedPodNum,
+					},
+				},
+			},
+		},
+	}
+
+	indexer := podInformer.Informer().GetIndexer()
+	// Add 2 pods with labels matching group
+	for i := 0; i < int(expectedPodNum); i++ {
 		pod := &corev1.Pod{
 			ObjectMeta: metav1.ObjectMeta{
 				Namespace: ns,
+				Name:      fmt.Sprintf("pod-%d", i),
 				Labels: map[string]string{
-					v1alpha1.GroupNameLabelKey: groupName,
+					workloadv1alpha1.GroupNameLabelKey: groupName,
 				},
-				Name: "pod1",
 			},
 		}
-		c := newTestControllerWithPods([]*corev1.Pod{pod}, ns)
-		if !c.isInferGroupOutdated(group, mi.Namespace, newHash) {
-			t.Errorf("expected true when pod has no revision label")
-		}
-	})
+		err := indexer.Add(pod)
+		assert.NoError(t, err)
+		store.AddRunningPodToInferGroup(utils.GetNamespaceName(mi), groupName, pod.Name, newHash)
+	}
 
-	t.Run("revision not match", func(t *testing.T) {
-		pod := &corev1.Pod{
-			ObjectMeta: metav1.ObjectMeta{
-				Namespace: ns,
-				Labels: map[string]string{
-					v1alpha1.GroupNameLabelKey: groupName,
-					v1alpha1.RevisionLabelKey:  "oldhash",
-				},
-				Name: "pod2",
-			},
-		}
-		c := newTestControllerWithPods([]*corev1.Pod{pod}, ns)
-		if !c.isInferGroupOutdated(group, mi.Namespace, newHash) {
-			t.Errorf("expected true when revision not match")
-		}
+	// Waiting for pod cache to sync
+	sync := waitForObjectInCache(t, 2*time.Second, func() bool {
+		pods, _ := controller.podsLister.Pods(ns).List(labels.SelectorFromSet(map[string]string{
+			workloadv1alpha1.GroupNameLabelKey: groupName,
+		}))
+		return len(pods) == int(expectedPodNum)
 	})
+	assert.True(t, sync, "Pods should be found in cache")
 
-	t.Run("revision match", func(t *testing.T) {
-		pod := &corev1.Pod{
-			ObjectMeta: metav1.ObjectMeta{
-				Namespace: ns,
-				Labels: map[string]string{
-					v1alpha1.GroupNameLabelKey: groupName,
-					v1alpha1.RevisionLabelKey:  newHash,
-				},
-				Name: "pod3",
+	fmt.Println(store.GetInferGroupByModelInfer(utils.GetNamespaceName(mi)))
+	ok, err := controller.checkInferGroupReady(mi, groupName)
+	assert.NoError(t, err)
+	assert.True(t, ok)
+
+	// case2: Pod quantity mismatch
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: ns,
+			Name:      "pod-1",
+			Labels: map[string]string{
+				workloadv1alpha1.GroupNameLabelKey: groupName,
 			},
-		}
-		c := newTestControllerWithPods([]*corev1.Pod{pod}, ns)
-		if c.isInferGroupOutdated(group, mi.Namespace, newHash) {
-			t.Errorf("expected false when revision matches")
-		}
+		},
+	}
+	err = indexer.Delete(pod)
+	assert.NoError(t, err)
+	sync = waitForObjectInCache(t, 2*time.Second, func() bool {
+		pods, _ := controller.podsLister.Pods(ns).List(labels.SelectorFromSet(map[string]string{
+			workloadv1alpha1.GroupNameLabelKey: groupName,
+		}))
+		return len(pods) == int(expectedPodNum)-1
 	})
+	assert.True(t, sync, "Pods should be found in cache after deletion")
+	store.DeleteRunningPodFromInferGroup(utils.GetNamespaceName(mi), groupName, "pod-1")
+
+	ok, err = controller.checkInferGroupReady(mi, groupName)
+	assert.NoError(t, err)
+	assert.False(t, ok)
+}
+
+// waitForObjectInCache waits for a specific object to appear in the cache
+func waitForObjectInCache(t *testing.T, timeout time.Duration, checkFunc func() bool) bool {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			t.Logf("Object not found in cache after %v timeout", timeout)
+			return false
+		case <-ticker.C:
+			if checkFunc() {
+				return true
+			}
+		}
+	}
 }

--- a/pkg/infer-controller/datastore/store.go
+++ b/pkg/infer-controller/datastore/store.go
@@ -69,10 +69,10 @@ const (
 
 var ErrInferGroupNotFound = errors.New("infer group not found")
 
-func New() (Store, error) {
+func New() Store {
 	return &store{
 		inferGroup: make(map[types.NamespacedName]map[string]*InferGroup),
-	}, nil
+	}
 }
 
 // GetInferGroupByModelInfer returns the list of inferGroups and errors

